File: Top-Interview-Questions-Leetcode/Easy-Collections/linked-list/merge-two-sorted-lists.cpp


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* merge_list = nullptr;
        ListNode* head = nullptr;
        
        while (list1 or list2) {
            if (list2 == nullptr) {
                if (merge_list) merge_list->next = list1, merge_list = merge_list->next;
                else merge_list = list1;
                list1 = list1->next;
            }
            else if (list1 == nullptr) {
                if (merge_list) merge_list->next = list2, merge_list = merge_list->next;
                else merge_list = list2;
                list2 = list2->next;
            }
            
            else if (list1->val < list2->val) {
                if (merge_list) merge_list->next = list1, merge_list = merge_list->next;
                else merge_list = list1;
                list1 = list1->next;
            }
            else {
                if (merge_list) merge_list->next = list2, merge_list = merge_list->next;
                else merge_list = list2;
                list2 = list2->next;
            }
            
            if (head == nullptr) head = merge_list;
        }
 
        return head;
    }
};

File: Top-Interview-Questions-Leetcode/Easy-Collections/linked-list/merge-two-sorted-lists.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        merge_list = None
        head = None
        
        while list1 != None and list2 != None:
            if list1.val < list2.val:
                if merge_list != None:
                    merge_list.next = list1
                    merge_list = merge_list.next
                else:
                    merge_list = list1
                
                list1 = list1.next
            else:
                if merge_list != None:
                    merge_list.next = list2
                    merge_list = merge_list.next
                else:
                    merge_list = list2
                
                list2 = list2.next
            
            if head == None:
                head = merge_list
        
        while list1 != None:
            if merge_list != None:
                merge_list.next = list1
                merge_list = merge_list.next
            else:
                merge_list = list1
                head = merge_list
            
            list1 = list1.next
        
        while list2 != None:
            if merge_list != None:
                merge_list.next = list2
                merge_list = merge_list.next
            else:
                merge_list = list2
                head = merge_list
            
            list2 = list2.next
        
        return head